Toolset update: VS 2022 17.14 Preview 1

StephanTLavavej opened the pull request:
# :scroll: Changelog

- Code cleanups:
  * Removed compiler bug workarounds.
- Infrastructure improvements:
  * Updated dependencies.
    + Updated build compiler to VS 2022 17.14 Preview 1.
    + Updated Google Benchmark to 1.9.1.
    + Updated Python to 3.13.2.

# :gear: Commits

* VS 2022 17.14 Preview 1.
* Remove workaround for DevCom-10373274 "vcruntime nothrow array operator new falls back on the wrong function".
* Remove workaround for DevCom-10808176 "MSVC doesn't properly destroy a loop variable in constant evaluation".
  + However, `mdspan/index_operator.pass.cpp` is still failing due to a likely MSVC constexpr bug, "note: failure was caused by a read of a variable outside its lifetime".

# :phone: `static operator()` Commits

All of MSVC/Clang/EDG now support C++23 WG21-P1169R4 `static operator()`, so let's update the codebase:

* Remove `#ifdef __cpp_static_call_operator` from `P1169R4_static_call_operator`.
  + This uses `usual_latest_matrix.lst`.
* Guard `std::function` CTAD for static call operators with `_HAS_CXX23`.
  + There's a question of whether we should attempt to expand this CTAD downlevel, to match how static call operators are supported downlevel. It would probably work, even for CUDA (which I believe currently lacks support for static call operators, but also doesn't appear to reject the SFINAE syntax here on sight). However, `std::function` CTAD is a "nice to have" feature, and we aren't obligated to support it downlevel. Restricting it to C++23 is less risky, and completely avoids any question of disrupting CUDA (as CUDA 12.4 supports only up to C++20).
* Suppress MSVC Future Technology warnings C5280 and C5281.
  + Clang uses the same warning option with different messages; record both.
* Use static call operators downlevel, except for CUDA.
  + In C++14 mode, as of VS 2022 17.14 Preview 1 and Clang 19.1.1:
    - MSVC does NOT define the feature-test macro, but accepts the code (with a C5280 warning).
    - Clang DOES define the feature-test macro, and accepts the code (with a `-Wc++23-extensions` warning).
    - EDG does NOT define the feature-test macro, but accepts the code silently.
  + For Intel's compiler, which we don't support but which we do avoid gratuitously breaking, my experiments on Compiler Explorer indicate that they have Clang's behavior.
  + CUDA 12.4 was released in March 2024, and I suspect that their EDG version doesn't support this yet.
  + Additionally, split the macro into `_STATIC_LAMBDA` to work around VSO-2383148 "C++23 static lambdas aren't accepted by the old lambda processor". This means that they aren't accepted in C++14/17, without enabling `/permissive-` (strict mode) or `/Zc:lambda` (new processor). This also means that they aren't accepted in C++20/23 when `/Zc:lambda-` (old processor) is used. (Note: C++20/23 with `/permissive` (permissive mode) still uses the new processor, and therefore works.) Because none of this is Standard-mandated (it's a "nice to have" minor performance improvement), we should just avoid static lambdas with MSVC while the bug remains active.
    - This will be fixed in 17.14 Preview 3 by MSVC-PR-611685.
* Enable libcxx tests for MSVC.
  + However, `range_adaptor_closure.pass.cpp `is blocked by a compiler bug affecting ranges.

# :gear: More Commits

* Drop the `wmemchr()` optimization for `unreachable_sentinel_t`.
  + Fixes `GH_002431_byte_range_find_with_unreachable_sentinel`, which regressed either with VS 2022 17.14 Preview 1
  or the newer Win11 SDK 26100 that I had unintentionally installed briefly; I didn't care enough to find out what was the cause. (Our README currently recommends installing the older Win11 SDK 22621 because that's what the MS-internal build currently uses.)
* Google Benchmark 1.9.1.
  + No MS-internal changes needed.
* 2025-datacenter-azure-edition.
  + The extra crispy flavor is said to be more efficient than the original recipe. It doesn't seem to make a difference for us, but it seems ideal to use.
* Python 3.13.2.
* Send `create-1es-hosted-pool.ps1` output to a log file.
  + I'm updating the `.gitignore` to ignore `*.log` anywhere in the tree, not just the root. (We don't have any such files in the repo.)
  + This captures output (via appending to ` >> $LogFile`) that I was previously piping to ` | Out-Null` to reduce script noise. I've had some trouble with sporadic failures and this should make it easier to diagnose in the future.
* Fix the CUDA installation.
  + In #5247, I encountered a CUDA installer hang, #5282. While working around it, I didn't notice that `nvcc` was no longer being added to the path, so the test harness was just skipping that test. And I didn't notice that we need to install `cudart` for headers that it wants to include.
  + I don't know how to get the installer to update the path when installing selected subpackages (I tried the "VS integration" one and it wasn't sufficient, and trying different subsets is extremely time-consuming). I'm just going to manually update the path.
  + I've verified that all 6 configurations of `GH_000639_nvcc_include_all` are now passing instead of being skipped.
* New pool.
* Teach `print_failures.py` to understand XPASS in XML.
  + This is parsing:
    ```xml
    <testcase classname="libc++.std/utilities/template_bitset/bitset_members" name="left_shift_eq.pass.cpp" time="8.86">
        <failure><![CDATA[]]></failure>
    </testcase>
    ```
  + It handles XPASS in JSON just fine.
* Fix STL-ASan-CI failures by upgrading `bitset.members` tests from FAIL to SKIPPED.
  + These tests are stressing `constexpr` step limits. One started passing for ASan, which makes little sense. All of the tests in this directory are similar, so skip 'em.
* Expand `memchr`/`wmemchr` comments, fix `__std_find_trivial_unsized_2`.

:white_check_mark: [STL-ASan-CI passed.](https://dev.azure.com/vclibs/STL/_build/results?buildId=18252&view=results)

# :infinity: Commits Forever

* Work around the MSVC-internal toolset being 17.12 and not supporting static call operators.
* Add an MSVC-internal skip for a test with `ADDITIONAL_COMPILE_FLAGS(cl-style-warnings)`.
  + This test was otherwise re-enabled after VSO-2319111 was fixed.
* Cite CWG-1656 with our usual style, improve wrapping.
  + Thanks @frederick-vs-ja!
* Code review feedback: clarify `memchr`/`wmemchr` comments.
  + Thanks @davidmrdavid!

StephanTLavavej commented:
/azp run STL-ASan-CI

AlexGuteniev commented on stl/inc/xutility:
```diff
@@ -6202,28 +6202,22 @@ namespace ranges {
-                        _Result = reinterpret_cast<_Ptr_t>(const_cast<wchar_t*>(_CSTD wmemchr(
-                            reinterpret_cast<const wchar_t*>(_First_ptr), static_cast<wchar_t>(_Val), _Count)));
-                    }
+                    // C23 7.32.4.6.9 "The wmemchr generic function"/2 doesn't work with unreachable_sentinel_t.
```
This seems too indirect to me. Maybe link `memchr` behavior quote instead?
```c
// C23 7.27.5.2 "The memchr generic function"/2 as if reads the characters sequentially.
// wmemchr does not have to do that

AlexGuteniev replied:
The same incorrect assumption about `wmemchr` is also made in `vector_algorithms.cpp`, it is in a compat path for no longer existent vectorization.

StephanTLavavej replied:
I can update the comment and the `vector_algorithms.cpp` usage is super easy to fix - thanks, I totally missed that!

StephanTLavavej commented:
/azp run STL-ASan-CI

davidmrdavid reviewed (commented):
Thanks for this work! 🦖 
Left some nits. After reaching consensus, I should be ready to approve.

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -6216,7 +6216,9 @@ namespace ranges {
-                    // C23 7.32.4.6.9 "The wmemchr generic function"/2 doesn't work with unreachable_sentinel_t.
+                    // C23 7.27.5.2 "The memchr generic function"/2 says "The implementation shall behave as if
+                    // it reads the characters sequentially and stops as soon as a matching character is found."
+                    // C23 7.32.4.6.9 "The wmemchr generic function"/2 lacks such wording, so we don't use wmemchr().
```
Excellent suggestion, thanks. I think there's a bit of ambiguity in that phrasing - is it:

* { we don't use wmemchr() } to { avoid issues with unreachable_sentinel_t }
* we don't { use wmemchr() to avoid issues with unreachable_sentinel_t }

I'll go with "we don't use wmemchr(), avoiding issues with unreachable_sentinel_t".

StephanTLavavej commented on stl/src/vector_algorithms.cpp:
```diff
@@ -4481,12 +4481,15 @@ extern "C" {
 // TRANSITION, ABI: preserved for binary compatibility
 const void* __stdcall __std_find_trivial_unsized_2(const void* const _First, const uint16_t _Val) noexcept {
-    return wmemchr(static_cast<const wchar_t*>(_First), static_cast<wchar_t>(_Val), SIZE_MAX);
+    // C23 7.32.4.6.9 "The wmemchr generic function"/2 lacks such wording, so we don't use wmemchr().
```
It's a few lines above, but you have a good point that we usually don't expect comments to have *cross-function* context, and during maintenance functions can be reordered/modified/etc. I'll update.

StephanTLavavej commented:
Thanks! I've pushed a conflict-free merge with `main`, fixed issues found by a trial mirror to the MSVC-internal repo, and addressed CR feedback (all commits described at the bottom of the PR description).

StephanTLavavej commented:
I'm *speculatively* mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.
